File: workers/cad/flatcraft_cad/templates/enclosed_shelf.py
"""Enclosed shelf — настінна полиця з 4-сторонньою cross-розгорткою.

Phase 3.0 PR 7b (ADR-027). Pydantic-параметри (7a) + 3D builder (7b).
Unfold/DXF — у unfold.py і export/dxf.py відповідно. Server handler (`/export`
routing) + PDF + R3F + web studio — PR 7c.

Геометрія (див. enclosed-shelf.ts для діаграми):
- bottom: width_mm × depth_mm (центральна площина).
- back: width_mm × depth_mm (вертикальна стінка, UP-bend).
- left/right: depth_mm × depth_mm (квадратні бокові, UP-bend по Y-осі).
- front rib (опц.): width_mm × stiffening_rib.height_mm (UP-bend).

`depth_mm` — спільний параметр для висоти back AND розмірів бокових стінок
(інваріант ADR-027 Рішення 5).
"""

from typing import Literal

import cadquery as cq
from pydantic import BaseModel, ConfigDict, Field, field_validator

from flatcraft_cad.templates.base import BendSpec, Template

ALLOWED_INNER_RADIUS_MM: tuple[float, ...] = (1.0, 2.5, 4.0, 5.0)


class EnclosedShelfSidePerforation(BaseModel):
    """Опційна декоративна перфорація бокових стінок (квадратні отвори)."""

    model_config = ConfigDict(frozen=True, populate_by_name=True)

    hole_size_mm: float = Field(ge=3, le=20, description="Сторона квадратного отвору.")
    pitch_x_mm: float = Field(ge=10, le=100, description="Pitch між отворами по X.")
    pitch_y_mm: float = Field(ge=10, le=100, description="Pitch між отворами по Y.")
    margin_mm: float = Field(ge=5, le=50, description="Відступ від країв бокової стінки.")


class EnclosedShelfStiffeningRib(BaseModel):
    """Опційне ребро жорсткості — front lip bottom (UP-bend)."""

    model_config = ConfigDict(frozen=True, populate_by_name=True)

    height_mm: float = Field(ge=5, le=50, description="Висота ребра після гибу.")


class EnclosedShelfBuildParameters(BaseModel):
    """Параметри enclosed_shelf для CadQuery-builder'у (паритет з TS-схемою).

    `bends`: tuple з 3 або 4 BendSpec залежно від наявності stiffening_rib.
    Дефолт усіх — 'up' (enclosed-форма, на відміну від wall_shelf 'down').
    """

    model_config = ConfigDict(frozen=True, populate_by_name=True)

    width_mm: float = Field(ge=300, le=1000, description="Довжина полиці (= ширина bottom).")
    depth_mm: float = Field(
        ge=100,
        le=300,
        description="Глибина полиці (= висота back AND сторона квадратних боковин).",
    )
    bend_radius_mm: float = Field(description="Внутрішній радіус гиба.")
    bend_angle_deg: Literal[90] = Field(default=90, description="MVP: тільки 90°.")
    bends: tuple[BendSpec, ...] = Field(
        default=(BendSpec(direction="up"),) * 4,
        description="Напрям 3-4 гибів: [back, left, right, rib?].",
    )
    thickness_mm: float = Field(gt=0, le=10, description="Товщина листа.")
    side_perforation: EnclosedShelfSidePerforation | None = Field(
        default=None, description="Опційна перфорація боковин."
    )
    stiffening_rib: EnclosedShelfStiffeningRib | None = Field(
        default=None, description="Опційне 5-те ребро жорсткості (front-lip bottom)."
    )

    @field_validator("bend_radius_mm")
    @classmethod
    def _radius_in_allowed_set(cls, value: float) -> float:
        if value not in ALLOWED_INNER_RADIUS_MM:
            raise ValueError(
                f"bend_radius_mm must be one of {list(ALLOWED_INNER_RADIUS_MM)}; got {value}"
            )
        return value

    @field_validator("bends")
    @classmethod
    def _bends_count_3_or_4(cls, value: tuple[BendSpec, ...]) -> tuple[BendSpec, ...]:
        if len(value) not in (3, 4):
            raise ValueError(f"bends must contain 3 or 4 items; got {len(value)}")
        return value


def build_enclosed_shelf(params: EnclosedShelfBuildParameters) -> cq.Workplane:
    """Збирає 3D enclosed_shelf union'ом 4-5 box'ів (Z — vertical).

    Конвенція осей:
    - X: ширина полиці (left → right), x ∈ [0, width_mm].
    - Y: глибина полиці (front → back), y ∈ [0, depth_mm].
    - Z: vertical, +Z вгору.

    Сегменти:
    - bottom: x∈[0,w], y∈[0,d], z∈[0,t]. Лежить плашмя.
    - back: x∈[0,w], y∈[d-t,d], z∈[t,t+d]. Вертикальна стінка ззаду.
    - left: x∈[0,t], y∈[0,d], z∈[t,t+d]. Вертикальна стінка зліва.
    - right: x∈[w-t,w], y∈[0,d], z∈[t,t+d]. Вертикальна стінка справа.
    - rib (опц.): x∈[0,w], y∈[0,t], z∈[t,t+rh]. Front-edge ребро жорсткості.

    Side perforation НЕ вирізається у 3D — лише у DXF (як і в інших шаблонах).
    """
    t = params.thickness_mm
    w = params.width_mm
    d = params.depth_mm

    bottom = cq.Workplane("XY").box(w, d, t, centered=(False, False, False))
    back = cq.Workplane("XY").box(w, t, d, centered=(False, False, False)).translate((0, d - t, t))
    left = cq.Workplane("XY").box(t, d, d, centered=(False, False, False)).translate((0, 0, t))
    right = cq.Workplane("XY").box(t, d, d, centered=(False, False, False)).translate((w - t, 0, t))

    result = bottom.union(back).union(left).union(right)

    if params.stiffening_rib is not None:
        rh = params.stiffening_rib.height_mm
        rib = cq.Workplane("XY").box(w, t, rh, centered=(False, False, False)).translate((0, 0, t))
        result = result.union(rib)

    return result


class EnclosedShelfTemplate(Template[EnclosedShelfBuildParameters]):
    """Адаптер шаблону — slug збігається з seed.ts."""

    name = "enclosed_shelf"

    def build(self, params: EnclosedShelfBuildParameters) -> cq.Workplane:
        return build_enclosed_shelf(params)

File: packages/types/src/templates/enclosed-shelf.ts
/**
 * Enclosed shelf — настінна полиця з 4-сторонньою cross-розгорткою
 * (Phase 3.0 PR 7, ADR-027).
 *
 * Геометрія:
 *
 *                ┌──────────────┐
 *                │              │
 *                │     BACK     │  висота = depth_mm (квадратна стінка)
 *                │              │
 *   ┌────────────┼──────────────┼────────────┐
 *   │  LEFT      │              │   RIGHT    │
 *   │ (depth ×   │   BOTTOM     │  depth ×   │  бокові стінки — квадрати depth × depth
 *   │  depth)    │ (width ×     │  depth)    │
 *   │            │  depth)      │            │
 *   └────────────┼──────────────┼────────────┘
 *                │ FRONT_RIB?   │  опційне 5-те ребро
 *                └──────────────┘
 *
 * Cross-розгортка означає 3-4 bend lines уздовж 2 осей (back+rib — уздовж X,
 * sides — уздовж Y). На відміну від wall_shelf/l_bracket з лінійним 1D unfold,
 * unfold для enclosed_shelf 2D — UnfoldedEnclosedShelf зберігає координати
 * кожного сегменту окремо (`bend_lines: tuple[BendLine2D, ...]`). Реалізація
 * unfold + builder + DXF/PDF — PR 7b/7c.
 *
 * Параметри:
 *   - width_mm: довжина полиці (= ширина bottom). 300-1000, крок 10.
 *   - depth_mm: глибина полиці (= висота back AND розмір кожної боковини).
 *     100-300, крок 10. Квадратні бокові — інваріант (`side_w = side_h = depth_mm`).
 *
 * Опційні features:
 *   - side_perforation: декоративні квадратні отвори на боковинах
 *     (як perforated_panel_square — той же grid, але на 2 окремих площинах).
 *   - stiffening_rib: ребро жорсткості на front edge bottom — мала смужка,
 *     що згинається UP. Прибирає прогин полиці у середині.
 */
import { z } from "zod";

import { BendSpecSchema } from "./bends.js";

/**
 * Опційна декоративна перфорація бокових стінок. Якщо `null` — без отворів.
 * Геометрія сітки повторює perforated_panel_square (4-vertex LWPOLYLINE).
 */
export const EnclosedShelfSidePerforationSchema = z.object({
  hole_size_mm: z.number().min(3).max(20).describe("group:Перфорація сторін|label:Сторона □ (мм)"),
  pitch_x_mm: z.number().min(10).max(100).describe("group:Перфорація сторін|label:Pitch X (мм)"),
  pitch_y_mm: z.number().min(10).max(100).describe("group:Перфорація сторін|label:Pitch Y (мм)"),
  margin_mm: z
    .number()
    .min(5)
    .max(50)
    .describe("group:Перфорація сторін|label:Відступ від країв (мм)"),
});
export type EnclosedShelfSidePerforation = z.infer<typeof EnclosedShelfSidePerforationSchema>;

/**
 * Опційне ребро жорсткості — front lip bottom (UP-bend).
 * `height_mm` — висота ребра після гибу.
 */
export const EnclosedShelfStiffeningRibSchema = z.object({
  height_mm: z.number().min(5).max(50).describe("group:Ребро жорсткості|label:Висота ребра (мм)"),
});
export type EnclosedShelfStiffeningRib = z.infer<typeof EnclosedShelfStiffeningRibSchema>;

export const EnclosedShelfParametersSchema = z.object({
  /** Довжина полиці (= ширина bottom). 300-1000 мм, крок 10. */
  width_mm: z.number().min(300).max(1000).describe("group:Розміри|label:Ширина полиці (мм)"),
  /**
   * Глибина полиці. Одночасно: висота back, ширина/висота квадратних боковин.
   * 100-300 мм, крок 10.
   */
  depth_mm: z
    .number()
    .min(100)
    .max(300)
    .describe("group:Розміри|label:Глибина (= висота back, мм)"),
  bend_radius_mm: z
    .union([z.literal(1), z.literal(2.5), z.literal(4), z.literal(5)])
    .describe("group:Гиб|label:Внутрішній радіус (мм)"),
  bend_angle_deg: z.literal(90).describe("group:Гиб|label:Кут гиба (°)"),
  /**
   * Напрям згину на кожен з 3-4 гибів: [back, left, right, rib?].
   * Дефолт — всі 'up' (UP-bend для enclosed-форми, на відміну від wall_shelf).
   */
  bends: z
    .array(BendSpecSchema)
    .min(3)
    .max(4)
    .default([{ direction: "up" }, { direction: "up" }, { direction: "up" }, { direction: "up" }])
    .describe("group:Гиб|label:Напрями згину (back/left/right/rib)"),
  side_perforation: EnclosedShelfSidePerforationSchema.nullable()
    .default(null)
    .describe("group:Перфорація сторін|label:Декоративна перфорація"),
  stiffening_rib: EnclosedShelfStiffeningRibSchema.nullable()
    .default(null)
    .describe("group:Ребро жорсткості|label:Ребро жорсткості"),
});

export type EnclosedShelfParameters = z.infer<typeof EnclosedShelfParametersSchema>;

export const ENCLOSED_SHELF_DEFAULT_PARAMETERS: EnclosedShelfParameters = {
  width_mm: 600,
  depth_mm: 200,
  bend_radius_mm: 2.5,
  bend_angle_deg: 90,
  bends: [{ direction: "up" }, { direction: "up" }, { direction: "up" }, { direction: "up" }],
  side_perforation: null,
  stiffening_rib: null,
};
